### PR TITLE
[LoopUnroll] Consider simplified operands while retrieving TTI instruction cost

### PR DESCRIPTION
**Pull request by skachkov-llvm**
Motivating example: https://godbolt.org/z/WcM6x1YPx
Here clang doesn't unroll loop with -Os, despite the fact that it will produce smaller and faster code. The issue is that we estimate cost of GEP as 1 after unrolling:
```
Loop Unroll: F[bar] Loop %for.body.i
  Loop Size = 5
Starting LoopUnroll profitability analysis...
 Analyzing iteration 0
Adding cost of instruction (iteration 0):   store i32 %x, ptr %arrayidx.i, align 4, !tbaa !7
Adding cost of instruction (iteration 0):   %arrayidx.i = getelementptr inbounds i32, ptr @array, i64 %indvars.iv.i
 Analyzing iteration 1
Adding cost of instruction (iteration 1):   store i32 %x, ptr %arrayidx.i, align 4, !tbaa !7
Adding cost of instruction (iteration 1):   %arrayidx.i = getelementptr inbounds i32, ptr @array, i64 %indvars.iv.i
 Analyzing iteration 2
Adding cost of instruction (iteration 2):   store i32 %x, ptr %arrayidx.i, align 4, !tbaa !7
Adding cost of instruction (iteration 2):   %arrayidx.i = getelementptr inbounds i32, ptr @array, i64 %indvars.iv.i
 Analyzing iteration 3
Adding cost of instruction (iteration 3):   store i32 %x, ptr %arrayidx.i, align 4, !tbaa !7
Adding cost of instruction (iteration 3):   %arrayidx.i = getelementptr inbounds i32, ptr @array, i64 %indvars.iv.i
  Exceeded threshold.. exiting.
  UnrolledCost: 8, MaxUnrolledLoopSize: 6
  will not try to unroll partially because -unroll-allow-partial not given
```
However, the more precise cost estimation is zero, because after unrolling we will not have non-constant index %indvars.iv.i, but some known compile-time constant: {0, 1, 2, 3}, and such addressing can be folded in given target architecture (RISC-V). My suggestion is to explicitly pass expected operands into TargetTransformInfo::getInstructionCost using SimplifiedValues map (e.g. for first iteration the mapping is i64 %indvars.iv.i -> i64 0).

### CODE REVIEW

**Comment by skachkov-sc** [comment]
Sorry, missed one test that is affected by this change (didn't have AMDGPU backend in local build). @arsenm , could you please review this change? It looks like the cost of such GEPs:
```
%arrayidx.out = getelementptr inbounds float, ptr addrspace(3) %out, i32 %indvars.iv
```
is considered as zero, so it unrolls loop in test_func_addrspacecast_cost_nonfree.

**Comment by skachkov-sc** [comment]
Ping

**Comment by skachkov-sc** [comment]
Ping
<details>
<summary>Number of completely unrolled loops after this change on test-suite (-Os build)</summary>

```
Program                                                                       loop-unroll.NumCompletelyUnrolled              
                                                                              before                            after  diff  
     test-suite :: MicroBenchmarks/LCALS/SubsetCLambdaLoops/lcalsCLambda.test   0.00                             27.00   inf%
           test-suite :: MicroBenchmarks/LCALS/SubsetARawLoops/lcalsARaw.test   0.00                             27.00   inf%
      test-suite :: MultiSource/Benchmarks/mediabench/jpeg/jpeg-6a/cjpeg.test   0.00                              2.00   inf%
    test-suite :: MultiSource/Benchmarks/DOE-ProxyApps-C/miniGMG/miniGMG.test   0.00                              2.00   inf%
                      test-suite :: MultiSource/Benchmarks/nbench/nbench.test   0.00                              1.00   inf%
               test-suite :: MultiSource/Benchmarks/mafft/pairlocalalign.test   0.00                              1.00   inf%
           test-suite :: MicroBenchmarks/LCALS/SubsetCRawLoops/lcalsCRaw.test   0.00                             27.00   inf%
           test-suite :: MicroBenchmarks/LCALS/SubsetBRawLoops/lcalsBRaw.test   0.00                             27.00   inf%
     test-suite :: MicroBenchmarks/LCALS/SubsetBLambdaLoops/lcalsBLambda.test   0.00                             27.00   inf%
test-suite :: MultiSource/Benchmarks/MiBench/consumer-jpeg/consumer-jpeg.test   0.00                              2.00   inf%
     test-suite :: MicroBenchmarks/LCALS/SubsetALambdaLoops/lcalsALambda.test   0.00                             27.00   inf%
                        test-suite :: MultiSource/Benchmarks/Olden/bh/bh.test   0.00                              3.00   inf%
                test-suite :: External/SPEC/CINT2006/456.hmmer/456.hmmer.test   0.00                              2.00   inf%
               test-suite :: External/SPEC/CFP2006/453.povray/453.povray.test   0.00                              2.00   inf%
                      test-suite :: MultiSource/Benchmarks/Bullet/bullet.test   1.00                              6.00 500.0%
                 test-suite :: MultiSource/Applications/JM/ldecod/ldecod.test   1.00                              5.00 400.0%
                 test-suite :: MultiSource/Applications/JM/lencod/lencod.test   2.00                             10.00 400.0%
        test-suite :: MultiSource/Benchmarks/ASCI_Purple/SMG2000/smg2000.test   1.00                              4.00 300.0%
               test-suite :: External/SPEC/CFP2006/447.dealII/447.dealII.test  65.00                            252.00 287.7%
            test-suite :: External/SPEC/CINT2006/464.h264ref/464.h264ref.test   3.00                              9.00 200.0%
                  test-suite :: MultiSource/Benchmarks/Olden/power/power.test   3.00                              7.00 133.3%
              test-suite :: MultiSource/Benchmarks/tramp3d-v4/tramp3d-v4.test  16.00                             33.00 106.2%
test-suite :: MultiSource/Benchmarks/MiBench/consumer-lame/consumer-lame.test   1.00                              2.00 100.0%
                test-suite :: MultiSource/Benchmarks/7zip/7zip-benchmark.test   4.00                              7.00  75.0%
        test-suite :: External/SPEC/CINT2006/483.xalancbmk/483.xalancbmk.test   7.00                              9.00  28.6%
                    test-suite :: External/SPEC/CINT2006/403.gcc/403.gcc.test  15.00                             17.00  13.3%
```
</details>

File: llvm/test/Transforms/LoopUnroll/AMDGPU/unroll-cost-addrspacecast.ll [hunk 1]
```diff
@@ -50,7 +50,7 @@ for.end:
 }
 
 ; CHECK-LABEL: @test_func_addrspacecast_cost_nonfree(
-; CHECK: br i1 %exitcond
+; CHECK-NOT: br i1 %exitcond
```
Comment (nikic) [inline llvm/test/Transforms/LoopUnroll/AMDGPU/unroll-cost-addrspacecast.ll:53]: I think you should adjust the `-unroll-threshold=49` option in this test to preserve the old behavior. If I understood correctly, the extra cost saving applies to all functions in this file equally, so just adjusting the threshold should work.

Comment (skachkov-sc) [reply llvm/test/Transforms/LoopUnroll/AMDGPU/unroll-cost-addrspacecast.ll:53]: Unfortunately, with this change it's now impossbile to find threshold that keeps old behavour for all tests in this file. The savings in test_func_addrspacecast_cost_nonfree are higher (I assume that for AMDGPU target GEP with constant offset is cheaper in addrspace(3)), so there is no threshold where first 2 tests are unrolled and the last is not.

Comment (nikic) [reply llvm/test/Transforms/LoopUnroll/AMDGPU/unroll-cost-addrspacecast.ll:53]: @arsenm Any idea how to preserve the intent of this test?

Comment (arsenm) [reply llvm/test/Transforms/LoopUnroll/AMDGPU/unroll-cost-addrspacecast.ll:53]: I don't know have any great ideas. Either split the test file up, or add some filler operations to up the costs in other functions

Comment (skachkov-sc) [reply llvm/test/Transforms/LoopUnroll/AMDGPU/unroll-cost-addrspacecast.ll]: I've come up with the following solution: add some extra level of indirection to get GEP indices so they will not become constant after loop unrolling; this hides the changes introduced by this patch and preserves the behaviour of original test (with slight increasing of unroll threshold). This is done by loading the GEP index from some global array (the additional cost of this loading is the same for all tests).

**Review by arsenm — APPROVED** [review]

**Pull request merged**